adapters/src/main/kotlin/adapters/repository/MongoDeviceRepository.kt

```diff
@@ -0,0 +1,157 @@
+
+
+private val COLLECTION_NAME = "device"
+
+class MongoDeviceRepository(
+    private val mongoTemplate: MongoTemplate
+) : DeviceRepository {
+
+    private val logger = LoggerFactory.getLogger(MongoDeviceRepository::class.java)
+
+    override fun retrieve(deviceId: DeviceId): Either<RetrieveError, Device> =
+        try {
+            mongoTemplate.findById(
+                deviceId.value.toString(),
+                MongoDevice::class.java,
+                COLLECTION_NAME
+            )!!.toDomain().right()
+        } catch (e: Exception) {
+            logger.error("Error retrieving device $deviceId due to ", e)
+            RetrieveError.DeviceRetrieveError.left()
+        }
+
+    override fun retrieveAll(): Either<RetrieveError, List<Device>> =
+        try {
+            mongoTemplate
+                .find(Query(), MongoDevice::class.java, COLLECTION_NAME)
+                .map { it.toDomain() }
+                .right()
+        } catch (e: Exception) {
+            logger.error("Error retrieving thing list due to ", e)
+            RetrieveError.DeviceRetrieveError.left()
+        }
+
+    override fun updateThingStatus(deviceId: DeviceId, thingId: ThingId, newStatus: Status): Either<SwitchError, Unit> =
+        try {
+            val query = Query(
+                Criteria.where("_id").`is`(deviceId.value.toString())
+                    .and("things._id").`is`(thingId.value.toString())
+            )
+
+            val update = Update().set("things.$.management", ThingManagement(newStatus))
+
+            mongoTemplate.updateFirst(query, update, COLLECTION_NAME)
+
+            Unit.right()
+        } catch (e: Exception) {
+            logger.error("Error updating thing status due to ", e)
+            SwitchError.DeviceNotFound.left()
+        }
+
+    override fun removeThing(deviceId: DeviceId, thingId: ThingId): Either<RetrieveError, Unit> =
+        try {
+            val query = Query(Criteria.where("_id").`is`(deviceId.value.toString()))
+
+            val update = Update().pull("things", BasicDBObject("_id", thingId.value.toString()))
+
+            mongoTemplate.updateFirst(query, update, COLLECTION_NAME)
+
+            Unit.right()
+        } catch (e: Exception) {
+            logger.error("Error removing thing $thingId from device $deviceId due to ", e)
+            RetrieveError.DeviceRemoveError.left()
+        }
+
+    override fun addThing(deviceId: DeviceId, thing: Thing): Either<AddThingError, Unit> =
+        try {
+            val query = Query.query(Criteria.where("_id").`is`(deviceId.value.toString()))
+
+            val update = Update().push("things", thing.toMongoThing())
+
+            mongoTemplate.updateFirst(
+                query, update, MongoDevice::class.java, COLLECTION_NAME
+            )
+            Unit.right()
+        } catch (e: Exception) {
+            logger.error("Error adding the thing ${thing.id} due to ", e)
+            //TODO should it add a new device with the thing?
```

Comment:
In case of device not found the addThing should simply returns an error.. someone else should take care of this complexity